active/InputValidation/InputValidation.md

```diff
@@ -0,0 +1,572 @@
+    
+public class Person : ValidationWrapper
+{
+    [MinLength(4, ErrorMessage = "Name must be at least 4 characters")]
+    [MaxLength(6, ErrorMessage = "Name must be no more than 6 characters")]
+    public string Name
+    {	
+        get { return GetValue<string>("Name"); }
+        set { SetValue("Name", value); }
+    }
+}
+```
+
+#### Code – INotifyDataErrorInfo base class
+The NotifyDataErrorInfoBase class (app code) provides a reusable implementation of INotifyDataErrorInfo (new API).  It’s the base of ValidationWrapper, which is the base class for Person.
+``` csharp
+public class NotifyDataErrorInfoBase : INotifyDataErrorInfo
+{
+    protected readonly Dictionary<string, ObservableCollection<ValidationError>> Errors;
+
+    protected NotifyDataErrorInfoBase()
+    {
+        Errors = new Dictionary<string, List<string>>();
+    }
+
+    public bool HasErrors => Errors.Any();
+
+    public event EventHandler<DataErrorsChangedEventArgs> ErrorsChanged;
+
+    public ObservableCollection<ValidationError> GetErrors(string propertyName)
+    {
+        return propertyName != null && Errors.ContainsKey(propertyName)
+            ? Errors[propertyName]
+            : new ObservableCollection<ValidationError>();
+    }
+
+    protected virtual void OnErrorsChanged(string propertyName)
+    {
+        ErrorsChanged?.Invoke(this, new DataErrorsChangedEventArgs(propertyName));
+    }
+
+    protected void ClearErrors()
+    {
+        foreach (var propertyName in Errors.Keys.ToList())
+        {
+            Errors.Remove(propertyName);
+            OnErrorsChanged(propertyName);
+        }
+    }
+}
+```
+#### Code – ValidationWrapper (no new APIs)
+ValidationWrapper derives from the above NotifyDataErrorInfoBase, and provides GetValue/SetValue methods for storage of subclass’ properties.  The SetValue performs validation.  This also provides an implementation of INotifyPropertyChanged.
+
+``` csharp
+public class ValidationWrapper : NotifyDataErrorInfoBase
+{
+
+    #region get/set
+    protected T GetValue<T>(string propertyName)
+    {
+        if (string.IsNullOrEmpty(propertyName))
+        {
+            throw new ArgumentException("Invalid property name", propertyName);
+        }
+
+        object value;
+        value = default(T);
+
+        return (T)value;
+    }
+
+    // Validate() gets called from inside SetValue
+    protected void SetValue<T>(string propertyName, T value)
+    {
+        if (string.IsNullOrEmpty(propertyName))
+        {
+            throw new ArgumentException("Invalid property name", propertyName);
+        }
+
+        var current = GetValue<T>(propertyName);
+
+        if (!value.Equals(current))
+        {
+            PropertyChanged?.Invoke(
+                    this, new PropertyChangedEventArgs(propertyName));
+            Validate(propertyName);
+        }
+    }
+    #endregion
+
+    #region Populate errors (C# - DataAnnotations)
+    private Task Validate(string propertyName)
```

Comment:
The implementation of this method looks to have errors.
It clears all errors--which may come from multiple properties.
It then only adds errors for the current property.
If multiple properties had errors before validating a property, you'd end up with an error collection that doesn't contain all the errors.